src/wasm.cc

```diff
@@ -209,8 +173,19 @@ void WasmBase::getFunctions() {
 #undef _GET_ALIAS
 #undef _GET
 
-#define _GET_PROXY(_fn) wasm_vm_->getFunction("proxy_" #_fn, &_fn##_);
-#define _GET_PROXY_ABI(_fn, _abi) wasm_vm_->getFunction("proxy_" #_fn, &_fn##_abi##_);
+  // Try to point the capability to one of the module exports, if the capability has been allowed.
+#define _GET_PROXY(_fn)                                                                            \
+  if (capabilityAllowed("proxy_" #_fn)) {                                                          \
+    wasm_vm_->getFunction("proxy_" #_fn, &_fn##_);                                                 \
+  } else {                                                                                         \
+    _fn##_ = nullptr;                                                                              \
+  }
+#define _GET_PROXY_ABI(_fn, _abi)                                                                  \
+  if (capabilityAllowed("proxy_" #_fn)) {                                                          \
+    wasm_vm_->getFunction("proxy_" #_fn, &_fn##_abi##_);                                           \
+  } else {                                                                                         \
+    _fn##_abi##_ = nullptr;                                                                        \
+  }
   _GET_PROXY(validate_configuration);
```

Comment:
Shouldn't we use `FOR_ALL_MODULE_FUNCTIONS` here?

Reply:
Changed. I don't think this is strictly necessary since `FOR_ALL_MODULE_FUNCTIONS` isn't reused anywhere (unlike `FOR_ALL_HOST_FUNCTIONS`, which is used twice). But it does make `getFunctions()` and `registerCallbacks` consistent with each other.